### PR TITLE
Allow agent to monitor endpoint

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

closes https://github.com/elastic/elastic-agent/issues/4083

This is a large and fairly wide-ranging set of changes that allows agent to monitor and report usage metrics for endpoint.
Specifically: 

- adds code to the coordinator and runtime controllers to filter PID changes up to the config system
- Adds logic to the metrics input generator to add a `system/process` metricset to monitor endpoint
- Add an integration test
- Change the behavior of the `local` env var in mage, so `local=agentbeat` will only build agentbeat, then pull other binaries from upstream. Needed this for testing, but kept it because it seemed vaguely useful. However, we don't need to merge it.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

### CODE REVIEW

**Comment by fearful-symmetry** [comment]
CI may be various kinds of broken until the snapshot beats build are updated, taking out of draft since technically everything needed is merged.

**Comment by fearful-symmetry** [comment]
Blocked by https://github.com/elastic/beats/pull/39730

**Comment by pierrehilbert** [comment]
Looks like this is no more blocked @pchila @faec @blakerouse

**Comment by fearful-symmetry** [comment]
@pierrehilbert CI will probably fail until snapshots are updated, but other than that, should be ready.

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 1]
```diff
@@ -1032,6 +1035,15 @@ func (c *Coordinator) runLoopIteration(ctx context.Context) {
 
 	case c.heartbeatChan <- struct{}{}:
 
+	case <-c.servicePidUpdate:
+		c.logger.Infof("got pid service update, refreshing config")
+		err := c.refreshComponentModel(ctx)
```
Comment (cmacknz) [inline internal/pkg/agent/application/coordinator/coordinator.go:1040]: If endpoint crashloops, which I've seen happen, we are going to be doing this continuously.

I suspect this is expensive and we likely want to rate limit it.

Comment (cmacknz) [reply internal/pkg/agent/application/coordinator/coordinator.go:1040]: This will be particularly bad if multiple sub-processes were to enter crash loops, or just restart frequently.

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 2]
```diff
@@ -1032,6 +1035,15 @@ func (c *Coordinator) runLoopIteration(ctx context.Context) {
 
 	case c.heartbeatChan <- struct{}{}:
 
+	case <-c.servicePidUpdate:
```
Comment (cmacknz) [inline internal/pkg/agent/application/coordinator/coordinator.go:1038]: Service PID update is a bit of a misleading title, it implies it has to do with the service runtime specifically, but you are just generically pulling this out of the checkin observed so it isn't limited to that.

Remove service from the name to clarify this.

File: internal/pkg/agent/application/monitoring/v1_monitor.go [hunk 3]
```diff
@@ -908,6 +910,87 @@ func (b *BeatsMonitor) injectMetricsInput(cfg map[string]interface{}, componentI
 		},
 	}
 
+	// add system/process metrics for services that can't be monitored via json/beats metrics
+	// If there's a checkin PID and it contains the "endpoint" string, assume we want to monitor it
```
Comment (cmacknz) [inline internal/pkg/agent/application/monitoring/v1_monitor.go:914]: Instead of just "endpoint" can you look for inputs using the service runtime https://github.com/elastic/elastic-agent/blob/e401afe9d7b3ac9f5b2aef7999642f8194d9d243/specs/endpoint-security.spec.yml#L23

That will remove the hard coding of the input name here and generalize better.

File: magefile.go [hunk 4]
```diff
@@ -1077,101 +1077,126 @@ func collectPackageDependencies(platforms []string, packageVersion string, requi
...
+				targetFile := filepath.Join(targetPath, filepath.Base(f))
+				packagesCopied += 1
+				if err := sh.Copy(targetFile, f); err != nil {
+					panic(err)
 				}
 			}
 		}
-	} else {
-		archivePath = movePackagesToArchive(dropPath, requiredPackages)
+		// a very basic footcannon protector; if packages are missing and we need to rebuild them, check to see if those files were copied
```
Comment (cmacknz) [inline magefile.go:1194]: Can all of this be a separate PR with a description for what it fixes? Doesn't seem related to the endpoint change.

File: testing/integration/helpers.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build integration
+
+package integration
```
Comment (cmacknz) [inline testing/integration/helpers.go:7]: The filename should include endpoint or just be endpoint.go or something. I don't think you want this file to be a generic dumping ground for utility functions.

File: testing/integration/monitoring_endpoint_test.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+type EndpointMetricsMonRunner struct {
+	suite.Suite
+	info    *define.Info
+	fixture *atesting.Fixture
+}
+
+func TestEndpointAgentServiceMonitoring(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		Group: "fleet",
```
Comment (cmacknz) [inline testing/integration/monitoring_endpoint_test.go:34]: Use one of the predefined constants, like `Fleet`

File: testing/integration/monitoring_endpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+	require.Eventually(runner.T(),
+		func() bool { return agentAndEndpointAreHealthy(runner.T(), ctx, agentClient) },
+		time.Minute*3,
+		time.Second,
+		"Endpoint component or units are not healthy.",
+	)
+}
+
+func (runner *EndpointMetricsMonRunner) TestEndpointMetrics() {
```
Comment (cmacknz) [inline testing/integration/monitoring_endpoint_test.go:103]: Is there a reason this can't just be another test in endpoint_security_test.go as well? 

Extending on that, could this just be sub-test added to the existing endpoint installation test to avoid having to install agent again just for this?

These are unit tests, they don't have to be completely isolated into their own test cases.

File: testing/integration/monitoring_endpoint_test.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	require.NoError(runner.T(), err)
+
+	endpointID := "endpoint-default"
+	require.Eventually(runner.T(), func() bool {
+
+		query := genESQueryByBinary(agentStatus.Info.ID, endpointID)
+		res, err := estools.PerformQueryForRawQuery(ctx, query, "metrics-elastic_agent*", runner.info.ESClient)
+		require.NoError(runner.T(), err)
+		runner.T().Logf("Fetched metrics for %s, got %d hits", endpointID, res.Hits.Total.Value)
+		return res.Hits.Total.Value >= 1
```
Comment (cmacknz) [inline testing/integration/monitoring_endpoint_test.go:117]: You should probably verify the metrics aren't 0. This test would pass if you monitored nothing, I think this same flaw might exist in all of these tests if you copied what was done in the others.

File: dev-tools/mage/settings.go [hunk 9]
```diff
@@ -88,7 +88,7 @@ var (
 
 	Snapshot      bool
 	DevBuild      bool
-	ExternalBuild bool
+	ExternalBuild string
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply dev-tools/mage/settings.go:91]: Yeah, the mage changes were done purely to make testing easier.  I can remove them.

[the remark this answers is not in the document]
Comment (cmacknz) [reply dev-tools/mage/settings.go:91]: This one I think actually changes the way the flag is parsed, bool I think will accept 0, 1, true, and false as input but string only accepts the ones you explicitly check for.

File: internal/pkg/agent/application/coordinator/coordinator.go (hunk 1 above)
Comment (fearful-symmetry) [reply internal/pkg/agent/application/coordinator/coordinator.go:1040]: yeah, the debug line isn't supposed to be there, oops. Do you think the overall routine should have a rate limit of some kind? We do a PID comparison in `applyComponentState`, so we won't be hammering the update if we get the same value a bunch of times. 

If we want to handle a sort of crashloop scenario, (endpoint restarts frequently, each time coming up with a new PID) that could be interesting, since we'd still want to track the PID, so we can actually update the config once the crash loop stops or slows down?

File: internal/pkg/agent/application/monitoring/v1_monitor.go (hunk 3 above)
Comment (fearful-symmetry) [reply internal/pkg/agent/application/monitoring/v1_monitor.go:914]: So, just to clarify, it looks like there's a component to the input spec: 
https://github.com/elastic/elastic-agent/blob/6573c38318ced4906c2bab2e43c1f72a00d3f210/pkg/component/input_spec.go#L25

That's only populated by endpoint and other components that we'd need to separately catch the PID for?

Comment (cmacknz) [reply internal/pkg/agent/application/monitoring/v1_monitor.go:914]: Yes that means it is going to be installed as an independent service and not a sub-process.

File: internal/pkg/agent/application/coordinator/coordinator.go (hunk 1 above)
Comment (fearful-symmetry) [reply internal/pkg/agent/application/coordinator/coordinator.go:1039]: Hmm, I see, that makes sense. So it sounds like we need some kind of separate thread that operates on a delay or otherwise bundles the PID updates.

**Comment by fearful-symmetry** [comment]
One again, some field definitions in Fleet are missing. Will need to update the integration...

**Comment by fearful-symmetry** [comment]
Still tinkering with the tests...

**Comment by fearful-symmetry** [comment]
So, I ended up adding a decent bit of logic to the integration test, hopefully to better shake out any problems that might exist with the metric update logic.

File: testing/integration/monitoring_endpoint_test.go (hunk 7 above)
Comment (fearful-symmetry) [reply testing/integration/monitoring_endpoint_test.go:108]: So, I ended up making the test a bit more complex instead, to the point where it feels like it makes more sense to have it as its own test? You might disagree, though.

File: internal/pkg/agent/application/coordinator/coordinator_state.go [hunk 10]
```diff
@@ -135,16 +135,25 @@ func (c *Coordinator) refreshState() {
...
+				pidRequiresUpdate = true
+			}
 			c.state.Components[i] = state
 			found = true
 			break
 		}
 	}
 	if !found {
 		c.state.Components = append(c.state.Components, state)
+		if state.State.CheckinPid != 0 {
```
Comment (blakerouse) [inline internal/pkg/agent/application/coordinator/coordinator_state.go:154]: Does the component only send the PID once or only when it changes?

The reason I ask is it seems like to me at this point, every checkin would result in the 30 second ticker recomputing the component model.

File: internal/pkg/agent/application/coordinator/diagnostics_test.go [hunk 11]
```diff
@@ -488,6 +488,7 @@ log_level: "warning"
 components:
   - id: "comp-1"
     state:
+      checkinpid: 0
```
Comment (blakerouse) [inline internal/pkg/agent/application/coordinator/diagnostics_test.go:491]: Why is this prefixed with `checkin`? All of this information in the state comes from checkin, so I am confused on the prefix naming.

File: internal/pkg/agent/application/monitoring/v1_monitor.go [hunk 12]
```diff
@@ -908,6 +910,92 @@ func (b *BeatsMonitor) injectMetricsInput(cfg map[string]interface{}, componentI
 		},
 	}
 
+	for _, comp := range componentList {
+		logp.L().Infof("input component %s: %#v", comp.ID, comp.InputSpec.Spec.Service)
```
Comment (blakerouse) [inline internal/pkg/agent/application/monitoring/v1_monitor.go:914]: Using the global logger is not really recommened, I don't think this should be used in new code.

Comment (fearful-symmetry) [reply internal/pkg/agent/application/monitoring/v1_monitor.go:914]: ooops, debugging statement I kept in there...

File: internal/pkg/agent/application/coordinator/coordinator_state.go (hunk 10 above)
Comment (fearful-symmetry) [reply internal/pkg/agent/application/coordinator/coordinator_state.go:154]: It should only mark `pidRequiresUpdate` if the component has an updated PID, or it's a new component with a non-zero pid. I don't see any part in the logic where `pidRequiresUpdate` would always be true?

File: internal/pkg/agent/application/coordinator/diagnostics_test.go (hunk 11 above)
Comment (fearful-symmetry) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go:491]: Is everything in the `ComponentState` object the result of a checkin? I didn't think it was, and that's where that struct name is coming from.

**Review by blakerouse — APPROVED** [review]
Overall this looks good.

I would really like to see `checkinpid` to just be `pid`, but not a blocker.

**Review by faec — APPROVED** [review]
Some suggestions on atomic handling, otherwise looks good
